### ISSUE TITLE
Error when casting ragged list to an array

### ISSUE THREAD

**Issue by danielcrane** [numpy/numpy#19113]
When casting a ragged list of `np.ndarray` to an array via `np.array(..., dtype=object)`, if the first dimension of each of the arrays in the list is the same size, a `ValueError` is raised.

When using `np.array(..., dtype=object)` on a ragged list of `np.ndarray` where the first dimension of each of the lists is not equal, the code works as expected and produces an `np.ndarray` of `.shape` `(n,)` where `n` the length of the original list.

### Reproducing code example:

```python
import numpy as np
x = [np.empty((10, 5)), np.empty((9,8))]
print(np.array(x, dtype=object).shape) # This outputs (2,) as expected

x = [np.empty((10, 5)), np.empty((10,8))]
print(np.array(x, dtype=object).shape) # This raises a ValueError
```

### Error message:

```
      4 x = [np.empty((10, 5)), np.empty((10,8))]
----> 5 print(np.array(x, dtype=object).shape)

ValueError: could not broadcast input array from shape (10,5) into shape (10,)

```

### NumPy/Python version information:

1.20.2 3.8.5 (default, Sep  4 2020, 07:30:14) 
[GCC 7.3.0]

**Comment by seberg** [comment]
Just to be sure: NumPy 1.18.x has the same behaviour (1.19 had larger changes).

<details> <summary> You are probably well aware that the reason is that the shape is discovered as `(2, 10)` instead of just `(2,)`, since that is as far as the shapes match up: </summary>

```
In [1]: x = [np.empty((10, 5)), np.empty((9,8))]
In [2]: np.core._multiarray_umath._discover_array_parameters(x, dtype=np.dtype("O"))
Out[2]: (dtype('O'), (2,))

In [3]: x = [np.empty((10, 5)), np.empty((10,8))]
In [4]: np.core._multiarray_umath._discover_array_parameters(x, dtype=np.dtype("O"))
Out[4]: (dtype('O'), (2, 10))
```
(The second part of the result is the shape – that function exposes the discovery privately, but skips filling the result).

</details>

Note that there are two colliding expectations:
1. If the shapes matches exactly `x = [np.arange(3), np.arange(3)]` we will fully unpack the array. In this case, create a `(2, 3)` object array.
2. Your expectation: We do not "rip apart" an array. Either its elements become the result elements (point 1.), or it is stored "as is". It cannot happen that different parts of the array are stored in different elements.

I am not actually sure that supporting 2. is great.  Point 1 will always adds the caveat "unless shapes match up perfectly".  There have been thoughts of adding an `ndmax=` or `depth=` argment (e.g. gh-16667, maybe we should look into that again).

Note that you can already get what you are expecting using assignment:
```
arr = np.empty((2,), dtype=object)
arr[...] = x
```
(you have to know the result shape ahead of time, though).

---

There is some one *weird* case with "reverse broadcasting", we allow in a few places:
```
In [15]: x = [np.ones((1,)), np.ones((1, 1))]

In [16]: np.core._multiarray_umath._discover_array_parameters(x, dtype=np.dtype("O"))
Out[16]: (dtype('O'), (2, 1))

In [17]: np.array(x, dtype=object)[1, 0]
Out[17]: 1.0
```
Where the additional dimension of the second array "vanishes". I am willing to bet that someone will notice if we change it, but maybe not enough to worry about.

**Comment by danielcrane** [comment]
> You are probably well aware that the reason is that the shape is discovered as `(2, 10)` instead of just `(2,)`

Indeed yes.

> If the shapes matches exactly x = [np.arange(3), np.arange(3)] we will fully unpack the array. In this case, create a (2, 3) object array.

I think this case is a little different from the one I posted above, since in this case both of their sizes match exactly, and actually `np.array(x)` would work without an error in the case you posted.

In this example:

```python
import numpy as np
x = [np.empty((10, 5)), np.empty((9,8))]
print(np.array(x, dtype=object).shape) # This outputs (2,) as expected

x = [np.empty((10, 5)), np.empty((10,8))]
print(np.array(x, dtype=object).shape) # This raises a ValueError
```

I think that the output of the final line could conceivably be returned as an object array of shape either `(2,10)` or `(2,)`, however I don't personally think it's great that it instead raises a `ValueError`.

**Comment by mattip** [comment]
I don't think we should touch this. There is no good answer. Users going down this path should pre-allocate their array according to their preferences, and assign as they wish, since no matter what NumPy does it will be wrong for some users. @danielcrane can you do something like:
```
# pre-determine shape and dtype
ret = np.empty(shape, dtype=dtype)
ret[0] = a
ret[1] = b
```

See also [NEP 34](https://numpy.org/neps/nep-0034-infer-dtype-is-object.html) which probably does not go far enough to make numpy "dumber" rather than "smarter".

**Comment by danielcrane** [comment]
@mattip Yes indeed, I'm aware that pre-allocating the array is a way around this, I just thought I'd raise it in case there's some way to make the behaviour more consistent for other users going forward, since it's rather counter-intuitive (to me) that this command will work in all cases **except** for if all elements have the same size in the first few dimensions.

**Comment by danielcrane** [comment]
> See also [NEP 34](https://numpy.org/neps/nep-0034-infer-dtype-is-object.html) which probably does not go far enough to make numpy "dumber" rather than "smarter".

To me personally it'd be better to go further down this route and in the case of ragged lists **require** users to pre-allocate their own array, since the current behaviour (where it works in most cases, but not all) feels like a bit of a trap.

I can easily imagine that there are people relying on `np.array(..., dtype=object)` with a ragged list that aren't aware of this particular idiosyncrasy that I'm reporting in this issue, that will eventually get some data that returns a `ValueError` due to having the same first (or first few) dimension sizes.

**Comment by kYwzor** [comment]
I recently ran into this behavior. I've found a mention of it as far back 2017 (#9225) and it also seems to be the same issue described in #17224.

The fact that (as mentioned by @danielcrane) this pattern always seems to work as expected, except on very specific scenarios where some dimensions happen to match, makes for a very evil bug. In my use case, I have lists of variable length which contain `ndarray`s of variable shape, so this only fails when a very unlucky combination of these happens to occur. I agree that NEP 34 could go farther and completely disallow this pattern, since it seems easy to make mistakes using it.

Maybe we could then have an alternative array constructor for explicitly creating an object array which mimics a list of objects? As for why one wouldn't simply stick with a list, I can give a (somewhat obscure) practical example where object arrays are explicitly required: [creating variable-length arrays](https://docs.astropy.org/en/stable/io/fits/usage/unfamiliar.html#creating-a-variable-length-array-table) for FITS binary tables using [Astropy](https://github.com/astropy/astropy).

Either way, for now this seems to be a decent general workaround (based on @mattip's answer) assuming we have a list `x` which contains `ndarray`s (or any object, really):
```
array = np.empty(len(x), dtype=object)
array[:] = x
```
